packages/vm/src/instance.rs

```diff
@@ -181,13 +189,127 @@ where
+    }
+
+    #[test]
+    fn func_errors_for_non_existent_function() {
+        let instance = mock_instance(&CONTRACT);
+        let missing_function = "bar_foo345";
+        match instance.func::<(), ()>(missing_function) {
+            Err(Error::ResolveErr { source, .. }) => match source {
+                ResolveError::ExportNotFound { name } => assert_eq!(name, missing_function),
+                _ => panic!("found unexpected source error"),
+            },
+            Err(e) => panic!("unexpected error: {:?}", e),
+            Ok(_) => panic!("must not succeed"),
+        }
+    }
+
+    #[test]
+    fn allocate_deallocate_works() {
+        let mut instance = mock_instance(&CONTRACT);
+
+        let sizes: Vec<usize> = vec![
+            0,
+            4,
+            40,
+            400,
+            4 * KIB,
+            40 * KIB,
+            400 * KIB,
+            4 * MIB,
+            40 * MIB,
+            400 * MIB,
+        ];
+        for size in sizes.into_iter() {
+            let region_ptr = instance.allocate(size).expect("error allocating");
+            instance.deallocate(region_ptr).expect("error deallocating");
+        }
+    }
+
+    #[test]
+    fn write_and_read_memory_works() {
+        let mut instance = mock_instance(&CONTRACT);
+
+        let sizes: Vec<usize> = vec![
+            0,
+            4,
+            40,
+            400,
+            4 * KIB,
+            40 * KIB,
+            400 * KIB,
+            4 * MIB,
+            // disabled for performance reasons, but pass as well
+            // 40 * MIB,
+            // 400 * MIB,
+        ];
+        for size in sizes.into_iter() {
+            let region_ptr = instance.allocate(size).expect("error allocating");
+            let original = vec![0u8; size];
+            instance
+                .write_memory(region_ptr, &original)
+                .expect("error writing");
+            let data = instance
+                .read_memory(region_ptr, size)
+                .expect("error reading");
+            assert_eq!(data, original);
+            instance.deallocate(region_ptr).expect("error deallocating");
+        }
+    }
+
+    #[test]
+    fn read_memory_errors_when_when_length_is_too_long() {
+        let length = 6;
+        let max_length = 5;
+        let mut instance = mock_instance(&CONTRACT);
+
+        // Allocate sets length to 0. Write some data to increase length.
+        let region_ptr = instance.allocate(length).expect("error allocating");
+        let data = vec![0u8; length];
+        instance
+            .write_memory(region_ptr, &data)
+            .expect("error writing");
+
+        match instance.read_memory(region_ptr, max_length) {
+            Err(Error::RegionLengthTooBigErr {
+                length, max_length, ..
+            }) => {
+                assert_eq!(length, 6);
+                assert_eq!(max_length, 5);
+            }
+            Err(err) => panic!("unexpected error: {:}", err),
```

Comment:
What is `{:}`?

I have seen `{}` for Display and `{:?}` for debug, but not this formatter

Reply:
:D I made that up today. And it compile. No idea what it does. Will check and probably migrate to `{:?}`

Reply:
`:` separates what for format (variable name goes before `:`) and how to format ([format_spec](https://doc.rust-lang.org/std/fmt/#syntax) comes after `:`). So this is most likely equivalent to `{}`.